### PR TITLE
Migrate index created version to IndexVersion

### PR DESCRIPTION
**Pull request by thecoop**
This is a large refactor that touches a lot of files, but unfortunately all the related search classes are all interconnected, with several associated test classes per query type.

I've highlighted everything but the most basic of changes with comments. Everything else is just a straight swap from `Version` to `IndexVersion`

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/analysis/PreConfiguredAnalysisComponent.java [hunk 1]
```diff
@@ -36,7 +36,7 @@ protected PreConfiguredAnalysisComponent(String name, PreBuiltCacheFactory.PreBu
 
     @Override
     public T get(IndexSettings indexSettings, Environment environment, String name, Settings settings) throws IOException {
-        Version versionCreated = indexSettings.getIndexVersionCreated();
+        Version versionCreated = indexSettings.getIndexVersionCreated().toVersion();
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/index/analysis/PreConfiguredAnalysisComponent.java:39]: Leaving the cache interface to another PR

File: server/src/main/java/org/elasticsearch/index/IndexModule.java [hunk 2]
```diff
@@ -339,7 +338,7 @@ public void addIndexOperationListener(IndexingOperationListener listener) {
      * @param name Name of the SimilarityProvider
      * @param similarity SimilarityProvider to register
      */
-    public void addSimilarity(String name, TriFunction<Settings, Version, ScriptService, Similarity> similarity) {
+    public void addSimilarity(String name, TriFunction<Settings, IndexVersion, ScriptService, Similarity> similarity) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/index/IndexModule.java:341]: This looks to be a plugin method, but we don't guarantee compatibility with plugins between versions. I don't know if this is likely to cause significant issues though, if we need to provide a backwards-compatible method for `Version`.

File: server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java [hunk 3]
```diff
@@ -53,13 +53,13 @@ public static Query newLenientFieldQuery(String field, RuntimeException e) {
         return Queries.newMatchNoDocsQuery("failed [" + field + "] query, caused by " + message);
     }
 
-    private static final Version NESTED_DOCS_IDENTIFIED_VIA_PRIMARY_TERMS_VERSION = Version.fromString("6.1.0");
+    private static final IndexVersion NESTED_DOCS_IDENTIFIED_VIA_PRIMARY_TERMS_VERSION = IndexVersion.fromId(6010099);
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java:56]: Do we still need this check, now 6.1.0 indices are not supported?

File: modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java [hunk 4]
```diff
@@ -282,7 +286,7 @@ Query percolateQuery(
             return new PercolateQuery(name, queryStore, documents, candidateQuery, searcher, filter, verifiedMatchesQuery);
         }
 
-        Tuple<BooleanQuery, Boolean> createCandidateQuery(IndexReader indexReader, Version indexVersion) throws IOException {
```
Comment (thecoop) [inline modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java:285]: Turns out this parameter wasn't actually used at all

File: server/src/main/java/org/elasticsearch/index/mapper/MapperRegistry.java [hunk 5]
```diff
@@ -70,14 +70,14 @@ public Map<String, RuntimeField.Parser> getRuntimeFieldParsers() {
      * Return a map of the meta mappers that have been registered. The
      * returned map uses the name of the field as a key.
      */
-    public Map<String, MetadataFieldMapper.TypeParser> getMetadataMapperParsers(Version indexCreatedVersion) {
-        if (indexCreatedVersion.onOrAfter(Version.V_8_0_0)) {
+    public Map<String, MetadataFieldMapper.TypeParser> getMetadataMapperParsers(IndexVersion indexCreatedVersion) {
+        if (indexCreatedVersion.onOrAfter(IndexVersion.V_8_0_0)) {
             return metadataMapperParsers;
-        } else if (indexCreatedVersion.major == 7) {
+        } else if (indexCreatedVersion.onOrAfter(IndexVersion.V_7_0_0)) {
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperRegistry.java:76]: IndexVersion doesn't have major/minor numbers, but the order of checks here ensures it still does the same thing

File: test/framework/src/main/java/org/elasticsearch/test/index/IndexVersionUtils.java [hunk 6]
```diff
@@ -112,4 +112,9 @@ public static IndexVersion getNextVersion(IndexVersion version) {
     public static IndexVersion randomCompatibleVersion(Random random) {
         return randomVersionBetween(random, IndexVersion.MINIMUM_COMPATIBLE, IndexVersion.CURRENT);
     }
+
+    /** Returns a random {@code IndexVersion} that is compatible with the previous version to {@code version} */
+    public static IndexVersion randomPreviousCompatibleVersion(Random random, IndexVersion version) {
```
Comment (thecoop) [inline test/framework/src/main/java/org/elasticsearch/test/index/IndexVersionUtils.java:117]: New test utils method

File: server/src/main/java/org/elasticsearch/index/mapper/MapperRegistry.java (hunk 5 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperRegistry.java:76]: thanks, this looks correct to me.

**Review by javanna — APPROVED** [review]
LGTM heavily based on trust :)

File: server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java (hunk 3 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/common/lucene/search/Queries.java:56]: We can look at this separately

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**